src/location/mod.rs

```diff
@@ -388,18 +434,41 @@ impl ObPartitionEntry {
         &self,
         part_id: i64,
     ) -> Option<&ObPartitionLocation> {
-        // logic_id = part_id in partition one
         self.parititon_location.get(&part_id)
     }
 
-    pub fn get_sub_partition_location_with_part_id(
+    pub fn get_partition_location_with_phy_id(
         &self,
-        part_id: i64,
+        phy_id: i64,
+        partid_tablet_map: Option<&HashMap<i64, i64>>,
+    ) -> Option<&ObPartitionLocation> {
+        // logic_id = part_id in partition one
+        let mut logic_id = phy_id;
+        if ob_vsn_major() >= 4 {
+            logic_id = partid_tablet_map.and_then(|m| m.get(&logic_id).copied())
+                .unwrap_or_else(|| {
```

Comment:
BTW, the `unwrap` here will lead to a panic which is unreasonable.

Reply:
unwrap_or_else will not panic, it will execute the closure if there is None or Err in the result instead.